[ONNX] Refactor MaxPool to support dynamic inputs

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113318

In https://github.com/pytorch/pytorch/pull/106270, the solution managed to solve the [`ceil_model` corner issue](https://github.com/onnx/onnx/issues/5711) with the usage of `get_pool_ceil_padding`. However, padding the ceil in converter side only works when we already know the input shapes, therefore, a regression happens when users want to do dynamic inputs.

This PR provides (1) refactor codes with torchlib implementation, (2) add dynamic shapes test, and (3) disable the corner tests with comments saying re-enable it when the [real fix from ONNX](https://github.com/onnx/onnx/pull/5741) is merged.

titaiwangms committed:
[ONNX] Fix MaxPool

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor MaxPool to support dynamic inputs"


In https://github.com/pytorch/pytorch/pull/106270, the solution managed to solve the [`ceil_model` corner issue](https://github.com/onnx/onnx/issues/5711) with the usage of `get_pool_ceil_padding`. However, padding the ceil in converter side only works when we already know the input shapes, therefore, a regression happens when users want to do dynamic inputs.

This PR provides (1) refactor codes with torchlib implementation, (2) add dynamic shapes test, and (3) disable the corner tests with comments saying re-enable it when the [real fix from ONNX](https://github.com/onnx/onnx/pull/5741) is merged.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/symbolic_opset10.py:
```diff
@@ -135,36 +135,164 @@ def topk(g: jit_utils.GraphContext, self, k, dim, largest, sorted, out=None):
+    if isinstance(dilation, int):
+        dilation = [dilation] * expand_size
+    else:
+        dilation = dilation  # type: ignore[assignment]
```
why do we need this no-op assignment? same below

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1454,13 +1454,42 @@ def test_maxpool_3d_ceil(self):
+            output_names=["output_0"],
+        )
+
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1454,13 +1454,42 @@ def test_maxpool_3d_ceil(self):
         x = torch.randn(1, 3, 32)
         self.run_test(model, x)
 
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1472,6 +1501,8 @@ def test_maxpool_2d_ceil_corner(self):
         x = torch.randn(1, 3, 32, 32)
         self.run_test(model, x)
 
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1484,6 +1515,8 @@ def test_maxpool_3d_ceil_corner(self):
         x = torch.randn(1, 3, 51, 52, 45)
         self.run_test(model, x)
 
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1492,6 +1525,8 @@ def test_maxpool_1d_ceil_corner_with_indices(self):
         x = torch.randn(1, 3, 32)
         self.run_test(model, x)
 
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1504,6 +1539,8 @@ def test_maxpool_2d_ceil_corner_with_indices(self):
         x = torch.randn(1, 3, 32, 32)
         self.run_test(model, x)
 
+    # Enable this when https://github.com/onnx/onnx/pull/5741 is supported by ONNX
```
```suggestion
    # TODO: Enable after https://github.com/onnx/onnx/pull/5741 or after ONNX 1.15.1+ is released
```

thiagocrepaldi reviewed (commented):
We can try to expedite https://github.com/onnx/onnx/pull/5741 and the ONNX bump on pytorch

thiagocrepaldi commented:
CI is failing with:

```bash
To execute this test, run the following from the base repo dir:
     python test_operators.py -k test_maxpool

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
================== 1 failed, 81 deselected, 2 rerun in 0.51s ===================

FINISHED PRINTING LOG FILE of onnx/test_operators 1/1 (/var/lib/jenkins/workspace/test/test-reports/onnx.test_operators_1.1_1mz_yii6_toprint.log)

onnx/test_operators 1/1 failed!
/var/lib/jenkins/workspace/tools/stats/upload_metrics.py:164: UserWarning: Error uploading metric td_experiment_1 to DynamoDB: Unsupported type "<class 'tools.testing.test_run.TestRun'>" for value "onnx/test_operators"
  warn(f"Error uploading metric {metric_name} to DynamoDB: {e}")
Emiting td_test_failure_stats
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1832, in <module>
    main()
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1779, in main
    run_tests(
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1607, in run_tests
    raise RuntimeError(
RuntimeError: onnx/test_operators 1/1 failed!
```

titaiwangms committed:
Update on "[ONNX] Refactor MaxPool to support dynamic inputs"


In https://github.com/pytorch/pytorch/pull/106270, the solution managed to solve the [`ceil_model` corner issue](https://github.com/onnx/onnx/issues/5711) with the usage of `get_pool_ceil_padding`. However, padding the ceil in converter side only works when we already know the input shapes, therefore, a regression happens when users want to do dynamic inputs.

This PR provides (1) refactor codes with torchlib implementation, (2) add dynamic shapes test, and (3) disable the corner tests with comments saying re-enable it when the [real fix from ONNX](https://github.com/onnx/onnx/pull/5741) is merged.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor MaxPool to support dynamic inputs"


In https://github.com/pytorch/pytorch/pull/106270, the solution managed to solve the [`ceil_model` corner issue](https://github.com/onnx/onnx/issues/5711) with the usage of `get_pool_ceil_padding`. However, padding the ceil in converter side only works when we already know the input shapes, therefore, a regression happens when users want to do dynamic inputs.

This PR provides (1) refactor codes with torchlib implementation, (2) add dynamic shapes test, and (3) disable the corner tests with comments saying re-enable it when the [real fix from ONNX](https://github.com/onnx/onnx/pull/5741) is merged.

[ghstack-poisoned]

titaiwangms commented:
@BowenBao @thiagocrepaldi @justinchuby PTAL

thiagocrepaldi reviewed (approved):
We will need to bump ONNX in the pytorch repro before re-enabling those tests, right?

titaiwangms commented:
@pytorchbot merge
